src/service/layers-impl.js

```diff
@@ -0,0 +1,876 @@
+    this.parentLayer_ = undefined;
+  }
+
+  /**
+   * Gets the current size of the element.
+   *
+   * @return {!SizeDef}
+   */
+  getSize() {
+    if (this.needsRemeasure_) {
+      this.remeasure();
+    }
+    return this.size_;
+  }
+
+  /**
+   * Gets current offset of the element relative to the parent layer, without
+   * taking scroll position into account.
+   *
+   * @return {!PositionDef}
+   */
+  getOffsetFromParent() {
+    if (this.needsRemeasure_) {
+      this.remeasure();
+    }
+    return this.position_;
+  }
+
+  /*
+   * Gets the current scrollTop of this layer.
+   *
+   * @return {number}
+   */
+  getScrollTop() {
+    this.updateScrollPosition_();
+    return this.scrollTop_;
+  }
+
+  /*
+   * Gets the current scrollLeft of this layer.
+   *
+   * @return {number}
+   */
+  getScrollLeft() {
+    this.updateScrollPosition_();
+    return this.scrollLeft_;
+  }
+
+  /**
+   * Returns the current scrolled position of the element relative to the layer
+   * represented by opt_parent (or opt_parent's parent layer, if it is not a
+   * layer itself). This takes into account the scrolled position of every
+   * layer in between.
+   *
+   * @param {Element=} opt_parent
+   * @return {!LayoutRectDef}
+   */
+  getScrolledPosition(opt_parent) {
+    // Compensate for the fact that the loop below will subtract the current
+    // scroll position of this element. But, this element's scroll position
+    // doesn't affect it's overall position, only its children.
+    // This is fine because the loop is guaranteed to roll at least once,
+    // zeroing the scroll.
+    let x = this.getScrollLeft();
+    let y = this.getScrollTop();
+
+    // Find the layer to stop measuring at. This is so that you can find the
+    // relative position of an element from some parent element, say the
+    // position of a slide inside a carousel, without any further measurements.
+    let stopAt = opt_parent ? LayoutElement.getParentLayer(opt_parent) : null;
+    for (let l = this; l !== stopAt; l = l.getParentLayer()) {
+      const position = l.getOffsetFromParent();
+      // Calculate the scrolled position. If the element has offset 200, and
+      // the element is scrolled 150, then the scrolled position is just 150.
+      // Note that the scrolled position shouldn't take into account the scroll
+      // position of this LayoutElement, so we've already compensated in
+      // declaring x and y.
+      x += position.left - l.getScrollLeft();
+      y += position.top - l.getScrollTop();
+    }
+
+    return positionLt(x, y);
+  }
+
+  /**
+   * Returns the absolute offset position of the element relative to the layer
+   * represented by opt_parent (or opt_parent's parent layer, if it is not a
+   * layer itself). This remains constant, regardless of the scrolled position
+   * of any layer in between.
+   *
+   * @param {Element=} opt_parent
+   * @return {!LayoutRectDef}
+   */
+  getOffsetPosition(opt_parent) {
```

Comment:
It may be clearer to move relative position calculation between layers into a separate class/utility entirely. All a "layer" needs is position/size/scroll, and finding relative positions shouldn't need access to layer internals.